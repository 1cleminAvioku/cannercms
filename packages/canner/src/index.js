export default from './components';
export {hocs} from './hocs';
export {default as Provder} from './components/Provider';
