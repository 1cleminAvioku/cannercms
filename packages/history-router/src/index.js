// @flow

import queryString from 'query-string';

type HistoryType = {
  location: {
    pathname: string,
    search: string
  },
  push: string => void
}

export default class HistoryRouter {
  constructor({baseUrl = '/', history}: {baseUrl: string, history: HistoryType}) {
    this.baseUrl = baseUrl;
    this.history = history;
  }

  getRoutes = () => {
    const {location: {pathname}} = this.history;
    const pathnameWithoutBaseUrl = pathname.substring(this.baseUrl.length);
    return deleteFirstSlash(pathnameWithoutBaseUrl).split('/');
  }

  getParams = () => {
    const {location: {search}} = this.history;
    const qs = queryString.parse(search);
    if (qs && qs.payload) {
      return {
        ...qs,
        payload: JSON.parse(qs.payload)
      }
    }
    return qs;
  }

  goTo = ({pathname, params = {}}: { pathname: string, params: Object}) => {
    const pathnameWithoutFirstSlash = deleteFirstSlash(pathname);
    const { payload, ...reset } = params;
    const qs = queryString.stringify({
      ...reset,
      payload: JSON.stringify(payload),
    });
    this.history.push(`${this.baseUrl}/${pathnameWithoutFirstSlash}${qs ? `?${qs}` : ''}`);
  }
}

function deleteFirstSlash(path) {
  if (path && path[0] === '/') {
    return path.substring(1);
  }
  return path;
}
