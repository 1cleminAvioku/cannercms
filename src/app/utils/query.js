/**
 * @flow
 */
type dataType = {
    totalPage: number,
    data: Array<any>
  } | Array<any>

import getByPath from 'lodash/get';

export default class Query {
  data: dataType

  constructor({data}: {data: Array<any>}) {
    this.data = data;
  }

  filter(filter: {[key: string]: any} = {}) {
    const data = this.getPureData();
    this.setData(Object.keys(filter).reduce((result, key) => {
      return this._filterByKey(result, key, filter[key]);
    }, data));
    return this;
  }

  sort(sort: {[key: string]: 1 | -1} = {}) {
    const data = this.getPureData();
    // eslint-disable-next-line max-len
    const collator = new Intl.Collator(undefined, {numeric: true, sensitivity: 'base', caseFirst: 'lower'});
    const key = Object.keys(sort)[0];
    const order = sort[key];
    this.setData(data.sort((a, b) => {
    // eslint-disable-next-line max-len
      return order > 0 ? collator.compare(a[key], b[key]) : collator.compare(b[key], a[key]);
    }));
    return this;
  }

  static turnPage(pagination?: {start: number, limit: number}) {
    if (!pagination) {
      return {page: 1, perPage: 10};
    }
    const {start, limit} = pagination;
    return {
      page: Math.ceil(start / limit) + 1,
      perPage: limit,
    };
  }

  page(pagination?: {page: number, perPage: number}) {
    const data = this.getPureData();
    if (!pagination) {
      return this;
    }
    const {page = 1, perPage = 10} = pagination;
    const queryData = data.slice(perPage * (page - 1), perPage * page);
    this.setData({
      goTo: (page) => ({
        start: (page - 1) * perPage,
        limit: perPage,
      }),
      page,
      totalPage: Math.ceil(data.length / perPage) || 0,
      data: queryData,
    });
    return this;
  }

  getPureData(): Array<any> {
    if (this.data && 'data' in this.data) {
      return (this.data: any).data;
    }
    return (this: any).data;
  }

  setData(data: dataType) {
    if (this.data && this.data.data) {
      (this.data: any).data = data;
    } else {
      this.data = data;
    }
  }

  getData() {
    return this.data;
  }

  // eslint-disable-next-line max-len
  _filterByKey(data: Array<any>, key: string, conditions: {[key: string]: any}) {
    return data.filter((datum) => {
      // eslint-disable-next-line max-len
      return Object.keys(conditions).reduce((result, conditionkey) => {
        // eslint-disable-next-line max-len
        return result && isValid(getByPath(datum, key.split('/'), ''), conditionkey, conditions[conditionkey]);
      }, true);
    });
  }
}

export function isValid(value: any, condition: string, conditionValue: any) {
  switch (condition) {
    case '$eq':
    case '$equal':
      // eslint-disable-next-line eqeqeq
      return value == conditionValue;
    case '$gt':
      return value > conditionValue;
    case '$gte':
      return value >= conditionValue;
    case '$lt':
      return value < conditionValue;
    case '$lte':
      return value <= conditionValue;
    case '$contains':
      return value.indexOf && value.indexOf(conditionValue) !== -1;
    case '$in':
      return conditionValue.indexOf && conditionValue.indexOf(value) !== -1;
    case '$regex':
      return new RegExp(conditionValue, 'g').test(value);
    default:
      return true;
  }
}

// eslint-disable-next-line max-len
export function passQuery(action: MutateAction, queryKey: string, totalPage: number) {
  if (action.type !== 'CREATE_ARRAY_ITEM') {
    return true;
  }
  // 用來判斷 CollectionUpadte 是否要在這個 query 的 data 下 mutate
  const allPage = queryKey === 'all';
  const query = allPage ? {filter: {}, pagination: {}} : JSON.parse(queryKey);
  const lastPage = getByPath(query, 'pagination.page', 1) === totalPage;
  const dataIsValid = new Query({data: [action.payload.value.toJS()]})
    .filter(query.filter).getPureData().length > 0;
  return dataIsValid && (allPage || lastPage);
}
