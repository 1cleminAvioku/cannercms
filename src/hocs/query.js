// @flow

import * as React from 'react';
import {Spin, Icon} from 'antd';
import type RefId from 'canner-ref-id';
import {Map, List, is} from 'immutable';
import Toolbar from './components/toolbar';
import type {Query} from '../query';
import {mapValues} from 'lodash';

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type Props = {
  type: string,
  items: Object,
  refId: RefId,
  query: Query,
  fetch: FetchDef,
  subscribe: SubscribeDef,
  updateQuery: Function,
  ui: string,
  path: string,
  pattern: string,
  relation: {
    type: string,
    to: string
  },
  toolbar: {
    sort?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    pagination?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    filter?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    toolbarLayout?: {
      component?: React.ComponentType<*>,
      [string]: *
    }
  }
};

type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    key: string;
    subscription: any;

    constructor(props: Props) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData();
    }

    UNSAFE_componentWillReceiveProps(props: Props) {
      const {refId} = this.props;
      
      if (refId.toString() !== props.refId.toString()) {
        this.queryData(props);
      } 
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    getRootValue = () => {
      return this.state.rootValue;
    }

    queryData = (props?: Props): Promise<*> => {
      const {refId, fetch} = props || this.props;
      return fetch(this.key).then(data => {
        this.setState({
          originRootValue: data,
          rootValue: parseConnectionToNormal(data),
          value: getValue(data, refId.getPathArr()),
          isFetching: false
        });
        this.subscribe();
      });
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe, refId} = this.props;
      const subscription = subscribe(this.key, (newOriginRootValue) => {
        const {originRootValue} = this.state
        if (shouldUpdate(originRootValue, newOriginRootValue)) {
          this.setState({
            originRootValue: newOriginRootValue,
            rootValue: parseConnectionToNormal(newOriginRootValue),
            value: getValue(newOriginRootValue, refId.getPathArr()),
          });
        }
      });
      this.subscription = subscription;
    }

    render() {
      const {value, isFetching, rootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, updateQuery, pattern} = this.props;
      if (isFetching) {
        return <Spin indicator={antIcon} />;
      }
      if (pattern === 'array' || (type === 'relation' && relation.type === 'toMany')) {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value.getIn(['edges'], new List()).map(item => item.get('node'))} />
        </Toolbar>;
      } else if (type === 'relation' && relation.type === 'toOne') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value || new Map()} />
        </Toolbar>;
      }
      return <Com {...this.props} rootValue={rootValue} value={value} />;
    }
  };
}

export function getValue(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (Map.isMap(result)) {
      if (result.has('edges') && result.has('pageInfo')) {
        return result.getIn(['edges', key, 'node']);
      }
      return result.get(key);

    } else if (List.isList(result)) {
      return result.get(key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: Map<string, *> | List<*>) {
  if (Map.isMap(value)) {
    value = ((value: any): Map<string, any>);
    if (value.has('edges') && value.has('pageInfo')) {
      return (value.get('edges'): any).map(edge => parseConnectionToNormal(edge.get('node')));
    }
    return value.map(item => parseConnectionToNormal(item));
  } else if (List.isList(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

function shouldUpdate(value: any, newValue: any) {
  return !is(value, newValue);
}