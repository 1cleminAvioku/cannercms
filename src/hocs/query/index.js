// @flow

import * as React from 'react';
import Sort from './sort';
import Pagination from './pagination';
import Filter from './filter';
import isString from 'lodash/isString';
import isUndefined from 'lodash/isUndefined';
import type RefId from 'canner-ref-id';

type Props = {
  filter: filterType,
  sort: sortType,
  params: Object,
  routes: Array<string>,
  componentId: string,
  fetch: FetchDef,
  refId: RefId
}

type State = {
  filter: filterType,
  sort: sortType,
  pagination: {
    start: number,
    limit: number,
  },
  paginationRes: any,
  value: any,
}

export default (Plugin: React.ComponentType<*>) => {
  return class PluginWithQuery extends React.PureComponent<Props, State> {
    constructor(props: Props) {
      super(props);
      this.state = {
        filter: {},
        sort: {},
        pagination: {
          start: 0,
          limit: 10,
        },
        paginationRes: {},
        value: null,
      };
      this.changeFilter = this.changeFilter.bind(this);
      this.changeSort = this.changeSort.bind(this);
      this.changePagination = this.changePagination.bind(this);
    }

    componentWillMount() {
      const {refId, subscribe, componentId} = this.props;
      this.queryData()
        .then(() => {
          subscribe(refId.toString(), componentId, 'value', (value) => {
            this.setState({
              value,
            });
          });
        });
    }

    queryData() {
      const {refId, fetch, componentId} = this.props;
      const {filter, sort, pagination} = this.state;
      return fetch(refId.toString(), componentId, {filter, sort, pagination}).then((ctx) => {
        this.setState({
          paginationRes: ctx.response.pagination,
          value: ctx.response.body,
        });
      });
    }

    changeFilter(condition) {
      this.setState({
        filter: condition,
      }, this.queryData);
    }

    changeSort(condition) {
      this.setState({
        sort: condition,
      }, this.queryData);
    }

    changePagination(start, limit) {
      this.setState({
        pagination: {
          start,
          limit,
        },
      }, this.queryData);
    }

    render() {
      const {filter, sort, params, routes} = this.props;
      const {paginationRes, pagination, value} = this.state;
      const op = params && params.op;
      if (!value) {
        return null;
      }
      return (
        <div>
          {
            // 如果是 string 則不是交給這個 filter 處理
            // 是 restful-qa 會在 render 前 filter
            // 目前只有 ireconcile 使用到
            op === 'create' || routes.length > 1 || isString(filter) || isUndefined(filter) ?
              null :
              <Filter onChange={this.changeFilter} schema={filter}/>
          }
          {
            op === 'create' || routes.length > 1 || isUndefined(sort) ?
              null :
              <Sort onChange={this.changeSort} options={sort}/>
          }
          <Plugin {...this.props} query={{filter: this.state.filter, sort: this.state.sort, pagination}} rootValue={value} value={value} showPagination={false}/>
          {
            op === 'create' || routes.length > 1 ?
              null :
              <Pagination onChange={this.changePagination} pagination={pagination} paginationRes={paginationRes}/>
          }
        </div>
      );
    }
  };
};

