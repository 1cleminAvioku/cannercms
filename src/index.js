export {hocs} from './hocs';
export {default as CMS} from './components';
export {default as Provder} from './components/Provider';
