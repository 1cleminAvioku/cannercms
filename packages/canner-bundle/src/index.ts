import webpack from 'webpack';
import WebpackDevServer from 'webpack-dev-server';
import {createConfig, CreateConfigArgsType, createWebConfig} from './utils/createWebpackConfig';
import transformSchemaToJson from './utils/transformSchemaToJSON';
import {SCHEMA_OUTPUT_PATH, SCHEMA_JSON_OUTPUT_PATH} from './config';
export function build(options?: CreateConfigArgsType) {
  const config = createConfig(options || {});
  return new Promise((resolve, reject) => {
    webpack(config)
      .run((err: any, stats) => {
        if (err) {
          console.error(err.stack || err);
          if (err.details) {
            return reject(err.details);
          }
        }
      
        const info = stats.toJson();
      
        if (stats.hasErrors()) {
          return reject(info.errors);
        }
        if (!options.webOnly) {
          const schemaPath = `${(options || {}).schemaOutputPath || SCHEMA_OUTPUT_PATH}/schema.node.js`;
          transformSchemaToJson(schemaPath, (options || {}).schemaJsonOutputPath || SCHEMA_JSON_OUTPUT_PATH);
        }
        resolve(stats);
      });
  });
}

export function serve(options?: CreateConfigArgsType) {
  const config = createWebConfig(options || {});
  const compiler = webpack(config);
  const devServerOptions = {...config.devServer};
  const server = new WebpackDevServer(compiler, devServerOptions);
  server.listen(config.devServer.port, '127.0.0.1', () => {
    console.log(`Starting server on http://localhost:${config.devServer.port}`);
  });
  return new Promise((resolve, reject) => {
    webpack(config)
      .watch({}, (err, stats) => {
        if (err || stats.hasErrors()) {
          return reject(err);
        }
        resolve(stats);
      });
  });
}
