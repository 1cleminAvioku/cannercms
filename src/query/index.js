// @flow

import {schemaToQueriesObject, objectToQueries} from './utils';
import {merge, get, set} from 'lodash';

export class Query {
  schema: Object
  queries: Object
  variables: Object

  constructor({schema}: {schema: Object}) {
    this.schema = schema;
    const {queriesObj, variables} = schemaToQueriesObject(schema, schema, {
      firstLayer: true
    });
    this.variables = variables;
    this.queries = queriesObj;
  }

  updateQueries = (pathArr: Array<string>, field: string, value: any) => {
    const path = `${pathArr.join('.fields.')}.${field}`;
    if (field === 'args') {
      const args = get(this.queries, path);
      merge(this.variables, Object.keys(args).reduce((result, key) => {
        result[args[key]] = value[key];
        return result;
      }, {}));
    } else {
      set(this.queries, path, value);
    }
  }

  getQueries = (pathArr: Array<string>): Object => {
    if (!pathArr || pathArr.length === 0) {
      return this.queries;
    }
    const path = pathArr.join('.fields.');
    return get(this.queries, path);
  }

  toGQL = (key?: string): string => {
    if (key) {
      const obj = this.getQueries([key]);
      return objectToQueries({[key]: obj});  
    } else {
      return objectToQueries(this.queries);
    }
  }

  getVairables = () => {
    return this.variables;
  }
}