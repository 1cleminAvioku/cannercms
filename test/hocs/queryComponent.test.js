import Toolbar, {parseOrder, parsePagination, parseWhere} from '../../src/hocs/components/toolbar';
import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import RefId from 'canner-ref-id';
import {Query} from '../../src/query';

Enzyme.configure({ adapter: new Adapter() });

describe('parse order', () => {
  it('should work', () => {
    const args = {
      orderBy: 'age_DESC'
    }
    const {orderField, orderType} = parseOrder(args.orderBy);
    expect(orderField).toBe('age');
    expect(orderType).toBe('DESC');
  });

  it('should return default', () => {
    const args = {

    };
    const {orderField, orderType} = parseOrder(args.orderBy);
    expect(orderField).toEqual(null);
    expect(orderType).toBe('ASC');
  });
});

describe('parse pagination', () => {
  it('should work', () => {
    const pagination = {
      first: 1,
      after: 3
    };

    expect(parsePagination(pagination)).toEqual(pagination);
  });
});

describe('parse where', () => {
  it('string 1', () => {
    const where = {
      name: 'name'
    };
    expect(parseWhere(where)).toEqual({
      name: {
        eq: 'name'
      }
    });
  });

  it('string 2', () => {
    const where = {
      name_eq: 'name'
    };
    expect(parseWhere(where)).toEqual({
      name: {
        eq: 'name'
      }
    });
  })

  it('number 1', () => {
    const where = {
      number: 1
    };
    expect(parseWhere(where)).toEqual({
      number: {
        eq: 1
      }
    });
  })

  it('number 2', () => {
    const where = {
      number_gt: 1
    };
    expect(parseWhere(where)).toEqual({
      number: {
        gt: 1
      }
    });
  });

  it('object 1', () => {
    const where = {
      author: {
        name: 'name'
      }
    };
    expect(parseWhere(where)).toEqual({
      author: {
        name: {
          eq: 'name'
        }
      }
    });
  })
});

describe('toolbar method', () => {
  let props, query, refId, items, value, toolbar;

  beforeEach(() => {
    const schema = {
      posts: {
        type: 'array',
        items: {
          type: 'object',
          items: {
            title: {
              type: 'string'
            },
            like: {
              type: 'number'
            },
            share: {
              type: 'number'
            }
          }
        }
      }
    };
    refId = new RefId('posts');
    query = new Query({
      schema
    });
    value = {
      edges: [{
        cursor: "0",
        node: {
          id: "0",
          title: "posts1"
        }
      }],
      pageInfo: {
        hasNextPage: true
      }
    }
    items = schema.posts.items.items;
    toolbar = {
      sort: {},
      pagination: {},
      filter: {}
    }
    props = {
      items,
      refId,
      query,
      value,
      toolbar,
      args: query.getQueries(refId.getPathArr()).args,
    };
  });

  it('should render', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should change order', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().changeOrder({
      orderField: 'like',
      orderType: 'DESC'
    });
    const {orderBy} = query.getQueries(refId.getPathArr()).args;
    expect(orderBy).toBe('RANDOM_KEY');
  });

  it('should change filter', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().changeFilter({
      share: {
        gt: 1
      }
    });
    const {where} = query.getQueries(refId.getPathArr()).args;
    expect(where).toEqual("RANDOM_KEY");
  });

  it('should nextPage', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().nextPage();
    const {pagination} = query.getQueries(refId.getPathArr()).args;
    expect(pagination).toEqual("RANDOM_KEY");
  });

  it('should prevPage', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().prevPage();
    const {pagination} = query.getQueries(refId.getPathArr()).args;
    expect(pagination).toEqual("RANDOM_KEY");
  });

  it('should changeSize', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().changeSize(20);
    const {pagination} = query.getQueries(refId.getPathArr()).args;
    expect(pagination).toEqual("RANDOM_KEY");
  });
});