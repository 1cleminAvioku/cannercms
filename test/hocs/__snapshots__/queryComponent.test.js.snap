// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toolbar method should render 1`] = `
<ToolbarLayout
  Filter={
    <Styled(FilterGroup)
      changeFilter={[Function]}
      items={
        Object {
          "like": Object {
            "type": "number",
          },
          "share": Object {
            "type": "number",
          },
          "title": Object {
            "type": "string",
          },
        }
      }
      where={undefined}
    />
  }
  Pagination={
    <PaginationPlugin
      changeSize={[Function]}
      hasNextPage={true}
      nextPage={[Function]}
      prevPage={[Function]}
      size={undefined}
    />
  }
  Sort={
    <Sort
      changeOrder={[Function]}
      items={
        Object {
          "like": Object {
            "type": "number",
          },
          "share": Object {
            "type": "number",
          },
          "title": Object {
            "type": "string",
          },
        }
      }
      options={undefined}
      orderField="RANDOM"
      orderType="ASC"
    />
  }
>
  <div />
</ToolbarLayout>
`;
